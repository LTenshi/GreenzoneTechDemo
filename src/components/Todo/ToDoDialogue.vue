<template>
	<v-card>
		<v-card-title>Add todo dialogue</v-card-title>
		<v-card-text>
			<v-btn-toggle v-model="isApiMode" dense>
				<v-btn :value="false">Client</v-btn>
				<v-btn :value="true">Api</v-btn>
			</v-btn-toggle>

			<!--
				This could be broken down into further 2 files, but since the file isn't too big I'll keep it here
				I found there were cases in the codebases I worked with where too much granurality slows down dev speed
			-->
			<div v-if="!isApiMode" class="dialogue-card-form-container">
				<v-form ref="todoForm" v-model="isFormInvalid">
					<v-text-field
						label="*Title"
						v-model="todoToAdd.title"
						:rules="titleRules"
					/>
					<!--The spec says that no empty content should exist, however I am assuming description is optional-->
					<v-textarea label="Description" v-model="todoToAdd.description" />
					<label for="file">Upload an image: </label>
					<input
						id="imageInput"
						class="mt-2 mb-2"
						type="file"
						name="file"
						accept="image/png, image/jpeg"
					/>
				</v-form>
				<div class="decision-handler">
					<v-btn class="mr-2" @click="closeDialog" color="error">Cancel</v-btn>
					<v-btn
						:disabled="!isFormInvalid"
						@click="addToDoItem"
						color="primary"
					>
						Save
					</v-btn>
				</div>
			</div>

			<div v-if="isApiMode" class="dialogue-card-form-container">
				<v-text-field
					type="number"
					label="Data number of todo"
					:min="1"
					v-model="selectedApiNumber"
				/>
				<div class="decision-handler">
					<v-btn class="mr-2" @click="closeDialog" color="error">Cancel</v-btn>
					<v-btn @click="addApiTodoItem" color="primary"> Save </v-btn>
				</div>
			</div>
		</v-card-text>
		<v-snackbar color="error" v-model="isSnackbarVisible" :timeout="2000">
			{{ errorText }}
		</v-snackbar>
	</v-card>
</template>

<script lang="ts">
import { InputValidationRule, InputValidationRules } from 'vuetify';
import { defineComponent } from 'vue';
import { ITodo } from '@/store/interfaces';
import store from '@/store';
import storeConstants from '@/store/storeConstants';
import { IApiPayload, convertToITodo } from './interfaces/apiPayload';

export default defineComponent({
	data: function () {
		return {
			todoToAdd: {
				title: '',
				description: '',
				id: 0,
				isCompleted: false,
				imageUrl: '',
			} as ITodo,
			isFormInvalid: false,
			isApiMode: false,
			isSnackbarVisible: false,
			selectedApiNumber: 1,
			errorText: '',
		};
	},
	computed: {
		//This was a big learning experience. "this" is not accessible in data section above
		titleRules(): InputValidationRules {
			const ruleEmpty: InputValidationRule = (value) =>
				!!value || 'Title is required';
			const ruleDuplication: InputValidationRule = (value) =>
				!store.getters[storeConstants.getters.IS_TITLE_DUPLICATED](value) ||
				'Title must be unique';
			const rules: InputValidationRules = [ruleEmpty, ruleDuplication];

			return rules;
		},
	},
	methods: {
		//There are quite a few methods here, one pattern I saw moved data/methods/computes/watches to their own files.
		//I keep an arbitiary number of 500 +/- 100 for max amount of lines in a vue file.
		//I find that's often more of a sweep under the rug move. Means the component is getting ultra bloated and should be broken down instead.
		//I think there are definetly cases where a separate file is a good idea e.g. large clientside algorithms if you definitely can't do them in API
		closeDialog(): void {
			this.$emit('closeDialog');
		},
		addToDoItem(): void {
			//Removed default check, since we already disable the save in dialog if invalid

			//Create a copy of the object, otherwise we get unwanted reactivity
			let payload = {} as ITodo;
			Object.assign(payload, this.todoToAdd);

			//Image handling
			const fileInput = document.getElementById(
				'imageInput'
			) as HTMLInputElement;

			if (fileInput && fileInput.files) {
				const selectedFile = fileInput.files[0];
				payload.imageUrl = URL.createObjectURL(selectedFile);
			}

			//Send off to the store
			store.dispatch(storeConstants.actions.ADD_TODO, payload);

			//Reset form
			const form = this.$refs.todoForm as HTMLFormElement;
			form.reset();
		},
		async fetchData(): Promise<IApiPayload> {
			try {
				return await (
					await fetch(
						`https://jsonplaceholder.typicode.com/todos/${this.selectedApiNumber}`
					)
				).json();
			} catch (err: Error | unknown) {
				console.log(err);
				this.isSnackbarVisible = true;
				this.errorText = 'Request to ToDo API failed';
				throw err;
			}
		},
		async addApiTodoItem(): Promise<void> {
			let payload = {} as ITodo;
			let response = await this.fetchData();

			payload = convertToITodo(response);
			if (
				store.getters[storeConstants.getters.IS_TITLE_DUPLICATED](payload.title)
			) {
				this.isSnackbarVisible = true;
				this.errorText = 'Requested ToDo already exsist! Try another one!';
				return;
			}
			store.dispatch(storeConstants.actions.ADD_TODO, payload);
		},
	},
});
</script>

<style lang="scss">
.dialogue-card-form-container {
	display: flex;
	flex-direction: column;
}

.decision-handler {
	align-self: end;
}
</style>
