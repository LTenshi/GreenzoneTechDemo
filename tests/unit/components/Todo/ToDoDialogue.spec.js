import ToDoDialogue from '../../../../src/components/Todo/ToDoDialogue.vue';
import Vuetify from 'vuetify';
import Vue from 'vue';
import { mount } from '@vue/test-utils';

//Fix for all vuetify related issues
Vue.use(Vuetify);

describe('Todo Dialogue Tests', () => {
	//Snapshot test - Make sure all the visible changes are the ones we expect
	test('Snapshot Test', () => {
		const wrapper = mount(ToDoDialogue);
		expect(wrapper.element).toMatchSnapshot();
	});

	test('by default the client form is shown', () => {
		const wrapper = mount(ToDoDialogue);

		expectIntitialFormCheckup(wrapper);
	});

	test('flick to the api form', async () => {
		const wrapper = mount(ToDoDialogue);
		const apiButton = wrapper.find('[id="btn-toggle-api"]');

		//Let's do our initial checkup just in case.
		expectIntitialFormCheckup(wrapper);
		//Press the button...
		apiButton.trigger('click');
		//...wait for the wrapper to update...
		await wrapper.vm.$nextTick();
		//...check that client form no longer exists...
		expect(wrapper.find('[id="container-client-form"]').exists()).toBe(false);
		//...but the api form does.
		expect(wrapper.find('[id="container-api-form"]').exists()).toBe(true);
	});

	test('flick to api form, then back to client', async () => {
		const wrapper = mount(ToDoDialogue);
		const clientButton = wrapper.find('[id="btn-toggle-client"]');
		const apiButton = wrapper.find('[id="btn-toggle-api"]');

		//Let's do our initial checkup just in case.
		expectIntitialFormCheckup(wrapper);
		//Press the button...
		apiButton.trigger('click');
		//...wait for the wrapper to update...
		await wrapper.vm.$nextTick();
		//...check that client form no longer exists...
		expect(wrapper.find('[id="container-client-form"]').exists()).toBe(false);
		//...but the api form does...
		expect(wrapper.find('[id="container-api-form"]').exists()).toBe(true);
		//...and click back on client...
		clientButton.trigger('click');
		//...wait for the wrapper to update...
		await wrapper.vm.$nextTick();
		//...check that api form no longer exists...
		expect(wrapper.find('[id="container-api-form"]').exists()).toBe(false);
		//...but the client form does...
		expect(wrapper.find('[id="container-client-form"]').exists()).toBe(true);
	});
});

function expectIntitialFormCheckup(wrapper) {
	//We are initially expecting for the client form to be shown.
	expect(wrapper.find('[id="btn-mode-toggle-group"]').exists()).toBe(true);
	//Make sure data is checking out...
	expect(wrapper.vm.$data.isApiMode).toBe(false);
	//...client form exists...
	expect(wrapper.find('[id="container-client-form"]').exists()).toBe(true);
	//...and api form doesn't.
	expect(wrapper.find('[id="container-api-form"]').exists()).toBe(false);
}
