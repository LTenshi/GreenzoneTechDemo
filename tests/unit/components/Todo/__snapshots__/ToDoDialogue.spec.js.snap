// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Dialogue Tests Snapshot Test 1`] = `
<div
  class="v-card v-sheet theme--light"
>
  <div
    class="v-card__title"
  >
    Add todo dialogue
  </div>
   
  <div
    class="v-card__text"
  >
    <div
      class="v-item-group theme--light v-btn-toggle v-btn-toggle--dense"
      id="btn-mode-toggle-group"
    >
      <button
        class="v-btn v-item--active v-btn--active v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
        id="btn-toggle-client"
        type="button"
        value="false"
      >
        <span
          class="v-btn__content"
        >
          Client
        </span>
      </button>
       
      <button
        class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
        id="btn-toggle-api"
        type="button"
        value="true"
      >
        <span
          class="v-btn__content"
        >
          Api
        </span>
      </button>
    </div>
     
    <div
      class="dialogue-card-form-container"
      id="container-client-form"
    >
      <form
        class="v-form"
        novalidate="novalidate"
      >
        <div
          class="v-input theme--light v-text-field"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-9"
                  style="right: 0px; position: absolute;"
                >
                  *Title
                </label>
                <input
                  id="input-9"
                  type="text"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <div
          class="v-input v-textarea theme--light v-text-field"
        >
          <div
            class="v-input__control"
          >
            <div
              class="v-input__slot"
            >
              <div
                class="v-text-field__slot"
              >
                <label
                  class="v-label theme--light"
                  for="input-12"
                  style="right: 0px; position: absolute;"
                >
                  Description
                </label>
                <textarea
                  id="input-12"
                  rows="5"
                />
              </div>
            </div>
            <div
              class="v-text-field__details"
            >
              <div
                class="v-messages theme--light"
              >
                <transition-group-stub
                  class="v-messages__wrapper"
                  name="message-transition"
                  tag="div"
                />
              </div>
            </div>
          </div>
        </div>
         
        <label
          for="file"
        >
          Upload an image: 
        </label>
         
        <input
          accept="image/png, image/jpeg"
          class="mt-2 mb-2"
          id="imageInput"
          name="file"
          type="file"
        />
      </form>
       
      <div
        class="decision-handler"
      >
        <button
          class="mr-2 v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default error"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            Cancel
          </span>
        </button>
         
        <button
          class="v-btn v-btn--disabled v-btn--has-bg theme--light v-size--default"
          disabled="disabled"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            
					Save
				
          </span>
        </button>
      </div>
    </div>
     
    <!---->
  </div>
   
  <div
    class="v-snack v-snack--bottom v-snack--has-background"
  >
    <transition-stub
      name="v-snack-transition"
    >
      <div
        class="v-snack__wrapper v-sheet theme--dark error"
        style="display: none;"
      >
        <div
          aria-live="polite"
          class="v-snack__content"
          role="status"
        >
          
		
	
        </div>
        <div
          class="v-snack__action "
        />
      </div>
    </transition-stub>
  </div>
</div>
`;
